### ISSUE TITLE
directory structure

### ISSUE THREAD

**Issue by wtholliday** [elalish/manifold#945]
I've created a Swift Package for Manifold so it can be easily used in Xcode: https://github.com/audulus/manifold/tree/spm2. Note the *spm2* branch.

I'd like to submit it as a PR, but I had to move files around because manifold has multiple `include/manifold` dirs which causes some confusion of Xcode when using the package and it can't find the main manifold header. Plus having to move files around makes it harder to keep my fork in sync.

What about having a single `include/manifold` directory which contains the headers?

**Comment by elalish** [comment]
Interesting! @pca006132 I think I asked a similar question when we were doing our recent directory restructuring, though I didn't know of any specific problems it would solve. This is definitely the right time to get our directory structure as clean as possible, since we're preparing for a v3.0 with some breaking changes anyway. Please open a PR and let's discuss there.

**Comment by pca006132** [comment]
I think having separate header directories for different module is fine. Not familiar with swift, but maybe you can add multiple targets in the package? E.g.

```

let package = Package(
    name: "manifold",
    platforms: [
        .macOS(.v11), .iOS(.v13)
    ],
    products: [
        .library(
            name: "manifold",
            targets: ["manifold"])
    ],
    dependencies: [
        .package(url: "https://github.com/audulus/glm", branch: "spm"),
        .package(url: "https://github.com/audulus/Clipper2", branch: "spm"),
    ],
    targets: [
        .target(
            name: "manifold",
            dependencies: ["manifold-utilities", "Clipper2"],
            path: "src",
            publicHeadersPath: "manifold/include"
        ),
        .target(
            name: "manifold-utilities",
            dependencies: ["glm"],
            publicHeadersPath: "utilities/include"
        )
        // ...
    ],
    cxxLanguageStandard: .cxx20
)
```

I guess there should be something that can generate this from cmake? Otherwise, you will just be mirroring our cmake config.

**Comment by stephomi** [comment]
Nope sorry, I do everything in C++ with a minimal amount of Objective C.
Mostly for cross-platform purposes, since Nomad is for [iPad](https://apps.apple.com/us/app/nomad-sculpt/id1519508653?mt=8&platform=ipad),  [Android](https://play.google.com/store/apps/details?id=com.stephaneginier.nomad&pcampaignid=pcampaignidMKT-Other-global-all-co-prtnr-py-PartBadge-Mar2515-1), and [Web](https://nomadsculpt.com/demo) (test version only).

For Manifold I was relying on the provided cmake but eventually I threw the towel because it was the one deps that was only breaking my build each time I updated it (glm/clipper2 dependencies colliding with my own (removing glm from the header would be nice), moving the manifold.a into the right folder, same for polygon.a, etc).
I guess I simply don't like (modern) cmake 😢.

So no I have my own Manifold cmake (thankfully the removal of thrust made it much easier).

```cmake
# -----------------------------------------------
target_link_libraries(nomad PRIVATE manifold)
target_compile_definitions(nomad PRIVATE HAS_MANIFOLD)
target_include_directories(nomad SYSTEM PRIVATE 
    lib/manifold/src/manifold/include
    lib/manifold/src/utilities/include)

# -----------------------------------------------
add_library(manifold OBJECT)
target_compile_options(manifold PRIVATE -w)

if(IS_DESKTOP)
    target_compile_options(manifold PRIVATE -DMANIFOLD_PAR='T')
    find_package(tbb REQUIRED)
    target_link_libraries(manifold PUBLIC TBB::tbb)
    # include("print_target_properties")
    # print_target_properties(TBB::tbb)
    target_include_directories(manifold PRIVATE Tbb::tbb)
elseif(IS_IOS)
    target_compile_options(manifold PRIVATE -DMANIFOLD_PAR='T')
    target_include_directories(manifold PRIVATE ${LIB_PATH}/include)
endif()

# headers
target_include_directories(manifold PRIVATE
    lib/glm
    lib/Clipper2/CPP/Clipper2Lib/include
    lib/manifold/src/manifold/include
    lib/manifold/src/collider/include
    lib/manifold/src/polygon/include
    lib/manifold/src/utilities/include)

# sources
target_sources(manifold PRIVATE
    lib/manifold/src/manifold/src/boolean3.cpp
    lib/manifold/src/manifold/src/boolean_result.cpp
    lib/manifold/src/manifold/src/constructors.cpp
    lib/manifold/src/manifold/src/csg_tree.cpp
    lib/manifold/src/manifold/src/edge_op.cpp
    lib/manifold/src/manifold/src/face_op.cpp
    lib/manifold/src/manifold/src/impl.cpp
    lib/manifold/src/manifold/src/manifold.cpp
    lib/manifold/src/manifold/src/properties.cpp
    lib/manifold/src/manifold/src/quickhull.cpp
    lib/manifold/src/manifold/src/smoothing.cpp
    lib/manifold/src/manifold/src/sort.cpp
    lib/manifold/src/manifold/src/subdivision.cpp
    lib/manifold/src/polygon/src/polygon.cpp)
```

**Comment by wtholliday** [comment]
> I guess there should be something that can generate this from cmake? Otherwise, you will just be mirroring our cmake config.

I don't know of something that will generate a `Package.swift` from CMake.

And yeah, probably can have multiple targets in the package and keep the separate directories. The main thing is just to avoid the (weird) multiple `include/manifold` directories. Only thing is that with that change, the actual files need to be changed, so instead of `#include "manifold/polygon.h"` it would be (probably) `#include "polygon/polygon.h"`

Kinda OT, but another thing that worries me about this swift package, is that I'm essentially just throwing the Manifold source files into my app without whatever configuration CMake does (compiler flags, etc). Would be nicer to be able to validate the build at least against the test suite. Which is why it might be preferable to generate an xcframework binary (even though apple discourages that) as I was trying to do here: https://github.com/elalish/manifold/issues/944

**Comment by pca006132** [comment]
I think the include files are fine? `manifold` is the name for the whole project, so we put headers inside `manifold/*`. The reason for having multiple `include/manifold` directories is that we want to split the headers into different modules, even though after install they should all be inside places like `/usr/include/manifold`.

**Comment by wtholliday** [comment]
I mean I think it's a bizarre decision, and I could use more colorful language, but you get the idea :) I fixed it in my fork at least.

**Comment by elalish** [comment]
@wtholliday This is mostly pretty recent, see #918. I made our original directory structure years ago when I knew pretty much nothing about C++ library packaging (and I still know little). Happy to change it to make it better - would you like to make a proposal in the form of a PR so our various users can comment on it?

**Comment by elalish** [comment]
@starseeker I'm looking at our directory structure now than #961 has landed - is there a reason the C bindings moved from `bindings/` to `src/`? Could we either move them back, or put them in a sub-directory like the other optional pieces?

**Comment by starseeker** [comment]
@elalish I took a stab at restoring it in https://github.com/starseeker/manifold/tree/restore_cbind - I'll make a PR to trigger the tests, but let me know if that's not what you're looking for.

**Comment by elalish** [comment]
Fixed by #961 and #967.
